### PR TITLE
Adding podman support and breaking up FDO init scripts

### PR DESCRIPTION
**Pull request by jkerry**
## Background

This change introduces podman as a deployment option and breaks up the init script templates to make things more readable for humans.

Relates OR Closes TF-12153

## How has this been tested?

This change will be tested extensively on all deployment / cloud combinations

### Did you add a new setting?

If no, you may delete these tasks.
If yes, please check each box after you have have added an issue in the TFE modules to add this setting:

- [ ] [AWS](https://github.com/hashicorp/terraform-aws-terraform-enterprise)
- [ ] [Azure](https://github.com/hashicorp/terraform-azurerm-terraform-enterprise)
- [ ] [GCP](https://github.com/hashicorp/terraform-google-terraform-enterprise)

## This PR makes me feel

optional gif describing your feelings about this pr

### CODE REVIEW

File: modules/runtime_container_engine_config/main.tf [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+        tmpfs = [
+          "/tmp:mode=01777",
+          "/run",
+          "/var/log/terraform-enterprise",
+        ]
+        ports = flatten([
+          "80:${var.http_port}",
+          "443:${var.https_port}",
+          local.active_active ? ["8201:8201"] : [],
+          var.metrics_enabled ? ["9090:9090"] : []
```
Comment (theoskolnik) [inline modules/runtime_container_engine_config/main.tf:62]: 💭 Consider adding 9091 for metrics HTTPS.

File: modules/tfe_init/main.tf [hunk 2]
```diff
@@ -7,9 +7,40 @@ locals {
...
+      ubuntu = {
+        docker = "${path.module}/templates/google.ubuntu.docker.tfe.sh.tpl",
+        podman = null
+      },
+      rhel = {
+        docker = "${path.module}/templates/google.rhel.docker.tfe.sh.tpl",
+        podman = null
+      }
+    }
+  }
```
Comment (theoskolnik) [inline modules/tfe_init/main.tf:41]: This is exactly the kind of separation we need here - well done! 🥳

File: modules/tfe_init/main.tf [hunk 3]
```diff
@@ -7,9 +7,40 @@ locals {
...
-
+  user_data_template = {
+    aws = {
+      ubuntu = {
+        docker = "${path.module}/templates/azurerm.ubuntu.docker.tfe.sh.tpl",
+        podman = null
+      },
+      rhel = {
+        docker = "${path.module}/templates/azurerm.rhel.docker.tfe.sh.tpl",
+        podman = "${path.module}/templates/azurerm.rhel.podman.tfe.sh.tpl"
```
Comment (theoskolnik) [inline modules/tfe_init/main.tf:18]: Little typo here, these should be `aws`.

Comment (theoskolnik) [reply modules/tfe_init/main.tf:18]: I also think it's fair to have RHEL version specific switches within the RHEL template file for each cloud (thinking about the difference between RHEL 8 and 9 for podman install).

File: modules/tfe_init/main.tf [hunk 4]
```diff
@@ -28,6 +59,7 @@ locals {
       tls_bootstrap_key_pathname  = local.tls_bootstrap_key_pathname
       tls_bootstrap_ca_pathname   = local.tls_bootstrap_ca_pathname
       compose                     = var.docker_compose_yaml
+      kube                        = var.podman_kube_yaml
```
Comment (theoskolnik) [inline modules/tfe_init/main.tf:62]: To be very explicit, we should probably call these vars `docker_compose` and `podman_kube_config` or something to that effect.

File: modules/tfe_init/templates/aws.rhel.podman.tfe.sh.tpl [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+%{ endif ~}
+
+%{ if enable_monitoring ~}
+install_monitoring_agents $log_pathname
+%{ endif ~}
+
+echo "[$(date +"%FT%T")] [Terraform Enterprise] Installing Podman" | tee -a $log_pathname
+
+dnf install -y container-tools
+systemctl enable --now podman.socket
```
Comment (theoskolnik) [inline modules/tfe_init/templates/aws.rhel.podman.tfe.sh.tpl:106]: 💭 Check for RHEL 8 vs 9 here?

File: modules/tfe_init/templates/aws.rhel.podman.tfe.sh.tpl [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+mount --options discard,defaults $device ${disk_path}
+chmod og+rw ${disk_path}
+echo "[Terraform Enterprise] Configuring automatic mounting of '$device' to directory at '${disk_path}' on reboot" | tee -a $log_pathname
+echo "UUID=$(lsblk --noheadings --output uuid $device) ${disk_path} ext4 discard,defaults 0 2" >> /etc/fstab
+%{ endif ~}
+
+%{ if enable_monitoring ~}
+install_monitoring_agents $log_pathname
+%{ endif ~}
+
```
Comment (theoskolnik) [inline modules/tfe_init/templates/aws.rhel.podman.tfe.sh.tpl:102]: Is all of this still really necessary?

**Review by theoskolnik — COMMENTED** [review]
I didn't take a deep look at every template file but I was thinking it would be a lot smaller for podman at least? Also have you run this code anywhere yet, does it actually bring up podman successfully? Maybe reference a branch build of this repo as a way to test?

File: modules/runtime_container_engine_config/main.tf (hunk 1 above)
Comment (jkerry) [reply modules/runtime_container_engine_config/main.tf:62]: Good catch. will do.

File: modules/tfe_init/templates/aws.rhel.podman.tfe.sh.tpl (hunk 6 above)
Comment (jkerry) [reply modules/tfe_init/templates/aws.rhel.podman.tfe.sh.tpl:102]: Mostly, yeah I think.  All of this is responsible for the various micro-configurations that we support and, if we let a bug out of the door with one of these, we'd end up with a patching scenario after GA.  Might not be needed for an individual release test case but it'll all come into play when we certify podman as a GA capability.

And it's already proven to work on these OSs, in these clouds.  There's little harm in standing on the shoulders of the giants of previous effort. If something ends up being straight up unnecessary we can cut it later.

**Comment by jkerry** [comment]
> I didn't take a deep look at every template file but I was thinking it would be a lot smaller for podman at least? Also have you run this code anywhere yet, does it actually bring up podman successfully? Maybe reference a branch build of this repo as a way to test?

kubectl validation passes on the output. I haven't branch targeted this yet but I totally will.  This PR is a draft until then

File: modules/tfe_init/main.tf (hunk 3 above)
Comment (theoskolnik) [reply modules/tfe_init/main.tf:18]: This definitely works: 

```
if grep -q -i "release 9" /etc/redhat-release
then
  dnf install -y container-tools
fi

if grep -q -i "release 8" /etc/redhat-release
then
  dnf module install -y container-tools
  dnf install -y podman-docker
fi
```

File: modules/runtime_container_engine_config/main.tf (hunk 1 above)
Comment (jkerry) [reply modules/runtime_container_engine_config/main.tf]: Addressed, used the configurable port numbers so it's not stuck on 9090/9091

File: modules/tfe_init/main.tf (hunk 3 above)
Comment (jkerry) [reply modules/tfe_init/main.tf]: I used this, but I make it an if/else with a fall through catch all.  That'll still fail in say rhel 7 but at least it's an attempt before bombing out.

**Review by theoskolnik — APPROVED** [review]
Adding here for context - John successfully tested podman module on mounted disk and external 🥳

**Pull request merged**